quickwit-pushapi/src/push_api_service.rs

```diff
@@ -18,28 +18,63 @@
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 use std::path::Path;
+use std::sync::Arc;
 
 use async_trait::async_trait;
 use quickwit_actors::{Actor, ActorContext, ActorExitStatus, ActorRunner, Handler, QueueCapacity};
+use quickwit_metastore::Metastore;
 use quickwit_proto::push_api::{
     CreateQueueRequest, DropQueueRequest, FetchRequest, FetchResponse, IngestRequest,
     SuggestTruncateRequest, TailRequest,
 };
 
-use crate::{iter_doc_payloads, Position, Queues};
+use crate::{iter_doc_payloads, Position, PushApiError, Queues};
 
 pub struct PushApiService {
     queues: Queues,
+    metastore: Arc<dyn Metastore>,
 }
 
 impl PushApiService {
-    pub fn with_queue_path(queue_path: &Path) -> crate::Result<Self> {
+    pub fn with_queue_path(
+        queue_path: &Path,
+        metastore: Arc<dyn Metastore>,
+    ) -> crate::Result<Self> {
         let queues = Queues::open(queue_path)?;
-        Ok(PushApiService { queues })
+        Ok(PushApiService { queues, metastore })
     }
 
-    fn ingest(&mut self, request: IngestRequest) -> crate::Result<()> {
+    async fn ingest(&mut self, request: IngestRequest) -> crate::Result<()> {
+        // Check all indexes exist.
+        let tasks = request.doc_batches.iter().map(|batch| {
+            let moved_metastore = self.metastore.clone();
+            let index_id = batch.index_id.clone();
+            tokio::spawn(async move {
+                let result = moved_metastore.check_index_available(&index_id).await;
+                (index_id, result)
+            })
+        });
+
+        let queue_ids = futures::future::try_join_all(tasks)
+            .await
+            .unwrap()
+            .into_iter()
+            .filter(|(_, result)| result.is_err())
+            .map(|(index_id, _)| index_id)
+            .collect::<Vec<_>>();
+        if !queue_ids.is_empty() {
+            return Err(PushApiError::IndexDoesNotExist { queue_ids });
+        }
+
         for doc_batch in &request.doc_batches {
+            // Attempt to create the queue if it does not exist.
+            if let Err(error) = self.queues.create_queue(&doc_batch.index_id) {
```

Comment:
I'm not fond of using error handling as logic. It makes the code harder to understand...

Could we use the following instead?

```rust
if !self.queues.exists() {
    self.queues.create_queue(&doc_batch.index_id)?;
}
```

Reply:
I was trying to avoid that extra call since we are in a loop. but you are right let's make it readable first.

Reply:
I was try to avoid that extra call to `self.queue_exists(queue_id)` inside `create_queue`, but you are right let's make it readable first